### DISCUSSION TITLE
Admin Only Route says bad token

### DISCUSSION THREAD

**Discussion by khrome83** [pocketbase/pocketbase#3406] [Q&A] [answered]
I am using the same code from the docs. 

```go
e.Router.GET("/hello", func(c echo.Context) error {
  return c.String(200, "Hello world!")
}, apis.ActivityLogger(app), apis.RequireAdminAuth())
```

When I visit this page, after logging into the application, I get the following response. 

```json
{
  "code": 401,
  "message": "The request requires valid admin authorization token to be set.",
  "data": {}
}
```

Is there another step to build a page that is SSR and Admin?

**Comment by ganigeorgiev** [comment]
In order to be able to access the above route you must send the request with `Authorization: ADMIN_TOKEN` header (_I'll consider removing the admin middleware from this particular example to avoid the confusion_).

> Is there another step to build a page that is SSR and Admin?

I'm not sure that I understand what you mean by this. Note that PocketBase authentication is fully stateless (aka. there are no sessions) and an admin/user is considered authenticated if a request has valid `Authorization` header.

**Reply by ganigeorgiev** [reply]
_I'm not sure if that's what you are looking for but you can also check https://github.com/pocketbase/pocketbase/discussions/989 for an example session cookie-like middleware._

**Reply by khrome83** [reply]
Ahh. I was kinda thinking that the login from `/_/` would start a session or cookie that could be reused on other pages. 

So I will need to have a admin login on another page. You don't normally set headers on basic get requests, as they are typically just browser links. So I am trying to figure out next steps. It's been a long time since I build a SPA. I have been doing a lot more server side rendered pages, where authentication is all server sessions across routes.

**Reply by ganigeorgiev** [reply, marked as answer]
> So I will need to have a admin login on another page.

Well not necessary. Similar to the example by @dschissler in https://github.com/pocketbase/pocketbase/discussions/989#discussioncomment-4161419, you can try to listen to the [`OnAdminAuthRequest`](https://pocketbase.io/docs/go-event-hooks/#onadminauthrequest) and set the cookie from there. Here is how this could look combined (_I haven't tested it so treat it as pseudo-code_):

```go
package main

import (
	"log"
	"net/http"

	"github.com/labstack/echo/v5"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/apis"
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/tokens"
	"github.com/pocketbase/pocketbase/tools/security"
	"github.com/spf13/cast"
)

const tokenCookieName = "pb_token"

func loadAuthContextFromCookie(app core.App) echo.MiddlewareFunc {
	return func(next echo.HandlerFunc) echo.HandlerFunc {
		return func(c echo.Context) error {
			tokenCookie, err := c.Request().Cookie(tokenCookieName)
			if err != nil || tokenCookie.Value == "" {
				return next(c) // no token cookie
			}

			token := tokenCookie.Value

			claims, _ := security.ParseUnverifiedJWT(token)
			tokenType := cast.ToString(claims["type"])

			switch tokenType {
			case tokens.TypeAdmin:
				admin, err := app.Dao().FindAdminByToken(
					token,
					app.Settings().AdminAuthToken.Secret,
				)
				if err == nil && admin != nil {
					// "authenticate" the admin
					c.Set(apis.ContextAdminKey, admin)
				}
			case tokens.TypeAuthRecord:
				record, err := app.Dao().FindAuthRecordByToken(
					token,
					app.Settings().RecordAuthToken.Secret,
				)
				if err == nil && record != nil {
					// "authenticate" the app user
					c.Set(apis.ContextAuthRecordKey, record)
				}
			}

			return next(c)
		}
	}
}

func main() {
	app := pocketbase.New()

	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		e.Router.Use(loadAuthContextFromCookie(app))

		e.Router.GET("/hello", func(c echo.Context) error {
			return c.String(200, "Hello world!")
		}, apis.ActivityLogger(app), apis.RequireAdminAuth())

		return nil
	})

	// auto set a cookie after admin authentication
	// 
	// (you can register similar hook for the auth records with `OnRecordAuthRequest`)
	app.OnAdminAuthRequest().Add(func(e *core.AdminAuthEvent) error {
		e.HttpContext.SetCookie(&http.Cookie{
			Name:     tokenCookieName,
			Value:    e.Token,
			Secure:   true,
			SameSite: http.SameSiteStrictMode,
			HttpOnly: true,
			MaxAge:   int(app.Settings().AdminAuthToken.Duration),
			Path:     "/",
		})
		return nil
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

_Note: if you want to be able to authenticate as admin and auth record at the same time you may need to use 2 separate cookies._

**Reply by dschissler** [reply]
For now I've decided not to use Pocketbase admin auth for my custom routes.  That is only for the PB Admin UI.  So I have another type of user admin that is a member of an admin group, the same as any other type of user.

**Reply by khrome83** [reply]
Makes sense @dschissler, I actually only need admin. For now. The website that actually reads the data isn't authenticated yet. So this is really just to manage back of house data.

**Comment by khrome83** [comment]
@ganigeorgiev this worked great.

Only comment I had was that there is no hook for `OnAdminAuthLogout` to catch the logout event and invalidate the cookie. But I don't think that is a deal breaker for my situation right now. 

This unblocks me for now, and I can improve in time. 

I get the stateless nature of Pocketbase. 

And I understand for front-end, not administrative, how that makes perfect sense. 
I wish there was a formal way to add more pages, even if it was not the same svelte framework. 

Like `/_/pages/:path` being adopted by the admin session (i know one does not exist), and adding a menu on the left the lets you mount links too it, and maybe maintains the left side bar, but lets the route swap out the non menu sidebar. 

That would give a lot of functionality to easily extend and grow the admin interface. 

For example, I am actually building a inventory management system for something custom. I have a need to create print labels (think address labels on a envelope) to sticker and track inventory. I can use a view to create the data in the way I want. But now I need a page that gives me access to a print stylesheet so I can send it too a inkjet. 

Not having the above, does not take anything away from the experience at all. But it seems like it would not be stretch to add this flexibility on day. Not advocating for modifying or manipulating the existing Svelte setup, outside of the menu to connect the pages to the main app. But even that could be raw html/js injected onto a route. 

Thank you for everything your doing, really great how well this is working.

**Reply by ganigeorgiev** [reply]
_Admin UI Add-ons API is in my todo to consider it (_you can watch https://github.com/pocketbase/pocketbase/issues/898_), but I want first to stabilize the internals to minimize introducing breaking changes to the eventual public plugin interface._
